finality-aleph/src/sync/service.rs

```diff
@@ -0,0 +1,266 @@
+            {
+                Ok(maybe_id) => maybe_id,
+                Err(e) => {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Error handling justification from {:?}: {}.", peer, e
+                    );
+                    return;
+                }
+            };
+            if let Some(block_id) = maybe_block_id {
+                if let Some(previous_block_id) = previous_block_id {
+                    self.backup_request(previous_block_id);
+                }
+                previous_block_id = Some(block_id);
+            }
+        }
+        if let Some(block_id) = previous_block_id {
+            self.request(block_id);
+        }
+    }
+
+    fn handle_request(&mut self, request: Request<J>, peer: N::PeerId) {
+        match self.handler.handle_request(request) {
+            Ok(action) => self.perform_sync_action(action, peer),
+            Err(e) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "Error handling request from {:?}: {}.", peer, e
+                );
+            }
+        }
+    }
+
+    fn handle_network_data(&mut self, data: NetworkData<J>, peer: N::PeerId) {
+        use NetworkData::*;
+        match data {
+            StateBroadcast(state) => self.handle_state(state, peer),
+            StateBroadcastResponse(justification, maybe_justification) => {
+                self.handle_justifications(
+                    iter::once(justification)
+                        .chain(maybe_justification)
+                        .collect(),
+                    peer,
+                );
+            }
+            Request(request) => {
+                let state = request.state().clone();
+                self.handle_request(request, peer.clone());
+                self.handle_state(state, peer)
+            }
+            RequestResponse(justifications) => self.handle_justifications(justifications, peer),
+        }
+    }
+
+    fn handle_task(&mut self, block_id: BlockIdFor<J>) {
+        use Interest::*;
+        match self.handler.block_state(&block_id) {
+            TopRequired {
+                know_most,
+                branch_knowledge,
+            } => {
+                self.send_request_for(block_id.clone(), branch_knowledge, know_most);
+                self.delayed_request(block_id);
+            }
+            Required {
+                know_most,
+                branch_knowledge,
+            } => {
+                self.send_request_for(block_id.clone(), branch_knowledge, know_most);
+                self.backup_request(block_id);
+            }
+            Uninterested => (),
+        }
+    }
+
+    fn handle_chain_event(&mut self, event: ChainStatusNotification<J::Header>) {
+        use ChainStatusNotification::*;
+        match event {
+            BlockImported(header) => {
+                if let Err(e) = self.handler.block_imported(header) {
```

Comment:
This error is converted to a log and ignored - how are we expected to recover from it?

Reply:
Ok, we're not, by design.